### PR TITLE
Fix equality check with tables and missing final variable assignment check with compound assignment

### LINKED ISSUE #29154: Equality check doesn't work for tables
== and != checks are disallowed for table values. Need to be allowed

### LINKED ISSUE #30300: Null point exception when check two tables are equivalent
**Description:**
$title. 

```ballerina
type Student record {
    readonly string name;
    int id;
};

function testSimpleSelectQueryWithTable() {
    table<Student> key(name) t = table [
        {name: "Amy", id: 1234},
        {name: "John", id: 4567}
    ];
    var ids = from var {id} in t select {id};
    table<record {|int id;|}> t2 = ids;
    assertEquality(table [{"id":1234},{"id":4567}], t2);
}

function assertEquality(anydata expected, anydata actual) {
    if expected == actual {
        return;
    }
    panic error("expected '" + expected.toString() + "', found '" + actual.toString () + "'");
}

```
```
[2021-05-03 10:26:13,238] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:2828)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangRecordLiteral.accept(BLangRecordLiteral.java:72)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeExpr(CodeAnalyzer.java:3750)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeExprs(CodeAnalyzer.java:3809)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:2809)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangTableConstructorExpr.accept(BLangTableConstructorExpr.java:67)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeExpr(CodeAnalyzer.java:3750)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeExprs(CodeAnalyzer.java:3809)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:3016)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangInvocation.accept(BLangInvocation.java:115)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeExpr(CodeAnalyzer.java:3750)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:2561)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangExpressionStmt.accept(BLangExpressionStmt.java:47)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:385)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:552)
	at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:58)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:385)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visitFunction(CodeAnalyzer.java:520)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:487)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeNode(CodeAnalyzer.java:385)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyzeTopLevelNodes(CodeAnalyzer.java:373)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.visit(CodeAnalyzer.java:365)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:167)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.CodeAnalyzer.analyze(CodeAnalyzer.java:354)
	at io.ballerina.projects.internal.CompilerPhaseRunner.codeAnalyze(CompilerPhaseRunner.java:225)
	at io.ballerina.projects.internal.CompilerPhaseRunner.performTypeCheckPhases(CompilerPhaseRunner.java:120)
	at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:383)
	at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
	at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:329)
	at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:176)
	at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:168)
	at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:97)
	at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:206)
	at io.ballerina.projects.Package.getCompilation(Package.java:131)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:68)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:166)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:58)
 
```

### LINKED ISSUE #30408: Final variable assignment check doesn't seems to run for compound assignment
**Description:**
From shell
```
=$ final int i = 0;                                                                                                                                                 took 1128 ms

=$ i += 1                                                                                                                                                            took 373 ms

=$ i                                                                                                                                                                 took 495 ms
1

=$  
```

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by lasinicl**
## Purpose
$title

Fixes #30300 NPE when check two tables are equivalent
Fixes #29154 Equality check doesn't work for tables
Fixes #30408 Final variable assignment check doesn't seems to run for compound assignment

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 1]
```diff
@@ -2753,7 +2758,27 @@ private static boolean isEqual(TableValueImpl lhsTable, TableValueImpl rhsTable,
             return true;
         }
 
-        return lhsTable.entrySet().equals(rhsTable.entrySet());
+        // check equality for keyless tables
+        if (!isLhsKeyedTable && !isRhsKeyedTable) {
```
Comment (pcnfernando) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:2762]: Here the order should be considered during the equality

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 2]
```diff
@@ -2741,19 +2740,24 @@ private static boolean isEqual(TableValueImpl lhsTable, TableValueImpl rhsTable,
...
+        boolean isLhsKeyedTable = ((BTableType) lhsTable.getType()).getFieldNames() != null &&
+                ((BTableType) lhsTable.getType()).getFieldNames().length > 0;
+        boolean isRhsKeyedTable = ((BTableType) rhsTable.getType()).getFieldNames() != null &&
+                ((BTableType) rhsTable.getType()).getFieldNames().length > 0;
+
+        Object[] lhsTableValues = lhsTable.values().toArray();
+        Object[] rhsTableValues = rhsTable.values().toArray();
+
+        if (isLhsKeyedTable == isRhsKeyedTable) {
+            for (int i = 0; i < lhsTableValues.length; i++) {
```
Comment (pcnfernando) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:2752]: If the table sizes are unequal, can return fast

Comment (lasinicl) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:2752]: Returns early at https://github.com/ballerina-platform/ballerina-lang/blob/master/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java#L2740

File: langlib/langlib-test/src/test/resources/test-src/valuelib_fromBalString_test.bal [hunk 3]
```diff
@@ -177,9 +177,6 @@ function testTableFromBalString() {
     if (tbl is table<any|error>) {
         tbl.add({ id: 3, name: "Jane", grade: 13 });
     }
```
Comment (pcnfernando) [inline langlib/langlib-test/src/test/resources/test-src/valuelib_fromBalString_test.bal:179]: Don't remove the test. Update/comment the issue and raise an issue for fromBalString() pointing here

**Review by pcnfernando — APPROVED** [review]
LGTM

File: langlib/langlib-test/src/test/resources/test-src/valuelib_fromBalString_test.bal (hunk 3 above)
Comment (lasinicl) [reply langlib/langlib-test/src/test/resources/test-src/valuelib_fromBalString_test.bal:179]: Created https://github.com/ballerina-platform/ballerina-lang/issues/30574

**Review by gimantha — APPROVED** [review]

**Pull request merged**